### PR TITLE
`Assessment`: Add a dropdown menu for structured grading instructions

### PR DESCRIPTION
**Pull request by valeriebucher**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

So far, the only way to use feedback from SGI is via drag-and-drop. However, this can lead to certain problems. Therefore, tutors now have the choice to also select the grading instructions via a drop-down menu.

### Description

The grading criteria of the exercise are forwarded to the feedback editor.  First the titles of the criteria are displayed in the drop-down menu, hovering brings up sub-menus that contain the instructions corresponding to the respective criterion. These are provided by a new additional component `textblock-feedback-dropdown`.

Some edge cases:
- if there is a criterion without a title: ‘(No title)’ is displayed
- if there is only one criterion: the instructions are displayed immediately within the dropdown menu
- if there are no criteria at all, the dropdown does not appear at all

### Steps for Testing

Prerequisites:

- 1 Tutor
- 1 Student submission of a text-exercise with SGI

1. Navigate to the Exercise Dashboard
2. Start the assessment
3. Select a text block and click on the button to see the dropdown.
4. If there is more than one criterion:
    4.1. Hover over the criterion title and click on the specific instruction.
    4.2. The text block should be marked with the feedback of the chosen instruction and the points should be calculated correctly.
5. If there is only one criterion:
    5.1. Check that there is no submenu, and instructions are shown immediately.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

https://user-images.githubusercontent.com/73841445/159497615-3305e731-47f3-4f35-b362-20447a453472.mov

If there is only one criterion:

Bildschirmfoto 2022-04-24 um 16 18 29

### CODE REVIEW

**Review by Hialus — APPROVED** [review]
Tested in testing session on TS1.
Works as described 👍🏻

**Review by b-fein — DISMISSED** [review]
Very nice feature! :+1: 

Code looks good, also worked without any issues during manual testing.

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts [hunk 1]
```diff
@@ -102,6 +107,22 @@ export class TextblockFeedbackEditorComponent implements AfterViewInit {
...
+        let colour;
+        if (instr.credits === 0) {
+            colour = '#fff2cc';
+        } else if (instr.credits < 0) {
+            colour = '#fbe5d6';
+        } else {
+            colour = '#e3f0da';
+        }
+        return colour;
+    }
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts:124]: I think get is the better prefix here since no attributes are changed when calling this method. I also think that these colors should be some constants (maybe there are even constants for this?).
```suggestion
    getInstrColour(instr: GradingInstruction): string | undefined {
        if (instr.credits === 0) {
            return '#fff2cc';
        } else if (instr.credits < 0) {
            return '#fbe5d6';
        } else {
            return '#e3f0da';
        }
        return undefined;
    }
```

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts [hunk 2]
```diff
@@ -177,6 +198,14 @@ export class TextblockFeedbackEditorComponent implements AfterViewInit {
         }
     }
 
+    updateAssessmentWithDropdown(instruction: GradingInstruction) {
+        this.feedback.gradingInstruction = instruction;
+        this.feedback.credits = instruction.credits;
+
+        this.feedback.correctionStatus = undefined;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts:205]: Why is this needed?

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html [hunk 3]
```diff
@@ -60,50 +60,113 @@
...
-                    (mouseenter)="mouseEnteredWarningLabel()"
-                    [translateValues]="{ affectedSubmissionsCount: textBlock.numberOfAffectedSubmissions }"
-                    >This feedback might affect <b>{{ textBlock.numberOfAffectedSubmissions }}</b> other similar submission/s.</span
-                >
+        <div class="input-group">
+            <div class="input-group-prepend">
+                <div *ngIf="criteria" class="dropdown" ngbDropdown>
+                    <button class="btn btn-primary dropdown-toggle" ngbDropdownToggle type="button"></button>
+                    <!--if there is only one criterion, dropdown-menu contains its instructions-->
+                    <ul *ngIf="criteria.length == 1" class="dropdown-submenu-single" id="test" ngbDropdownMenu>
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:68]:
```suggestion
                    <ul *ngIf="criteria.length === 1" class="dropdown-submenu-single" id="test" ngbDropdownMenu>
```

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html [hunk 4]
```diff
@@ -60,50 +60,113 @@
...
+                            <a (click)="updateAssessmentWithDropdown(instruction)" class="dropdown-item" [style.background-color]="setInstrColour(instruction)">
+                                <table class="table layout-table">
+                                    <tbody>
+                                        <tr>
+                                            <td class="td-score fw-bold">{{ instruction!.credits }}P</td>
+                                            <td>
+                                                <span class="td-instruction">{{ instruction!.instructionDescription }}</span>
+                                            </td>
+                                            <td class="td-usage-count">
+                                                <span *ngIf="instruction!.usageCount && instruction!.usageCount !== 0; else usageCountInfinity">{{ instruction!.usageCount }}</span>
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:79]: 0 is already falsy.  (same below)
```suggestion
                                                <span *ngIf="instruction!.usageCount; else usageCountInfinity">{{ instruction!.usageCount }}</span>
```

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html [hunk 5]
```diff
@@ -60,50 +60,113 @@
...
+                                                <ng-template #usageCountInfinity>&#8734;</ng-template>
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:80]: Maybe add a comment here explaining that these seemingly random numbers are the infinity symbol. (same below)

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html [hunk 6]
```diff
@@ -60,50 +60,113 @@
...
+                            <a class="dropdown-item" *ngIf="criterion.title; else withoutTitle"
+                                >{{ criterion.title }}&nbsp;<fa-icon class="float-end" [icon]="faAngleRight"></fa-icon
+                            ></a>
+                            <ng-template #withoutTitle
+                                ><a class="dropdown-item">(No title) <fa-icon class="float-end" [icon]="faAngleRight"></fa-icon></a
+                            ></ng-template>
+                            <ul class="dropdown-menu dropdown-submenu flex-column flex-wrap">
+                                <li *ngFor="let instruction of criterion!.structuredGradingInstructions">
+                                    <a (click)="updateAssessmentWithDropdown(instruction)" class="dropdown-item" id="item" [style.background-color]="setInstrColour(instruction)">
+                                        <table class="table layout-table">
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:101]: This table-section is duplicated from above, can this be avoided somehow?

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.scss [hunk 7]
```diff
@@ -89,3 +89,83 @@ label {
...
+        padding: 0.25rem 0.75rem;
+    }
+
+    li {
+        position: relative;
+
+        &:hover {
+            background-color: #e9ecef;
+        }
+        &:hover > .dropdown-submenu {
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.scss:120]: Codacy mentions multiple missing new lines in this file.

**Review by Strohgelaender — COMMENTED** [review]
I have a few questions and code suggestions:

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html (hunk 6 above)
Comment (valeriebucher) [reply src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:101]: I know :/ but not that I know of.

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts (hunk 1 above)
Comment (valeriebucher) [reply src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts:124]: Yes your suggestion with 'get' makes sense, I will adapt it. I think it's okay to leave the colors like this, there are no specific constants.

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts (hunk 2 above)
Comment (valeriebucher) [reply src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.ts:205]: This concerns the tutor training when they assess example submissions and receive feedback on what they have done wrong. The correction status (which can be correct or erroneous) must then be reset. But I'll add a comment to this.

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html (hunk 5 above)
Comment (akesfeden) [reply src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html:81]: As a small addition I can suggest `&infin;` as a somewhat self-documenting alternative but the comment works fine too.

File: src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html (hunk 6 above)
Comment (valeriebucher) [reply src/main/webapp/app/exercises/text/assess/textblock-feedback-editor/textblock-feedback-editor.component.html]: @Strohgelaender @Dominik-Weinzierl I updated this pr and provide the duplicated table-section within an additional component now.

**Review by cremertim — APPROVED** [review]
tested locally. Everything works as described in the testing steps. Awesome feature !! great work👍

**Review by Hialus — DISMISSED** [review]
Retested locally. Works as described.
Code also looks good to me. 👍🏻

**Review by filip-gregurevic — DISMISSED** [review]
Looks good to me and works as expected! 🚀

**Review by willergeorg — COMMENTED** [review]
Tested on TS3. Worked as expected in general. The only thing I noticed is a misaligned design if the grading instructions are very short and multiple criteria exist:
image

But one can argue that this is quite an edge case.

**Review by willergeorg — APPROVED** [review]
Tested again. Looks even better now, thank you!

**Review by max-bergmann — APPROVED** [review]
Tested on TS3. Works as expected – great feature.

**Review by filip-gregurevic — APPROVED** [review]
Re-approve 🚀

**Review by Hialus — APPROVED** [review]
Re-approve 👍🏻

**Review by jpbernius — APPROVED** [review]
Code looks good :+1:
A quick test was successful.

Let's ship :shipit:
